src/Castle.Core/DynamicProxy/Generators/MethodWithInvocationGenerator.cs

```diff
@@ -147,11 +157,19 @@ private Expression SetMethodInterceptors(ClassEmitter @class, INamingScope namin
 
 			var methodInterceptorsField = BuildMethodInterceptorsField(@class, MethodToOverride, namingScope);
 
+			Expression targetTypeExpression;
+			if (getTargetTypeExpression != null)
+			{
+				targetTypeExpression = getTargetTypeExpression(@class, MethodToOverride);
+			}
+			else
+			{
+				targetTypeExpression = new MethodInvocationExpression(null, TypeUtilMethods.GetTypeOrNull, getTargetExpression(@class, MethodToOverride));
+			}
+
 			var emptyInterceptors = new NewArrayExpression(0, typeof(IInterceptor));
 			var selectInterceptors = new MethodInvocationExpression(selector, InterceptorSelectorMethods.SelectInterceptors,
-			                                                        new MethodInvocationExpression(null,
-				                                                        TypeUtilMethods.GetTypeOrNull,
-				                                                        getTargetExpression(@class, MethodToOverride)),
+			                                                        targetTypeExpression,
```

Comment:
The simplest possible bugfix would be to wrap the result of `getTargetExpression(...)` with a call to `TypeUtil.GetTypeOrNull` only if it is not a `TypeTokenExpression` already. I discarded this simpler solution because it feels like a hack. (It requires out-of-band knowledge about the kind of `Expression` that class proxies will produce inside their version of `getTargetExpression`.) The fix proposed in this PR is more elaborate, but arguably cleaner.